pkg/global/constant_rate_trace_sampler.go

```diff
@@ -0,0 +1,86 @@
+package global
+
+import (
+	"sync"
+	"time"
+
+	"github.com/buildbarn/bb-storage/pkg/clock"
+
+	"go.opencensus.io/trace"
+)
+
+// ConstantRateTraceSampler uses a token bucket algorithm to sample traces. This is useful as an alternative
+// to probability-based sampling when the service has a low number of requests per second.
+type ConstantRateTraceSampler struct {
+	clock               clock.Clock
+	maxTokens           int64
+	periodDurationNanos int64
+	tokensPerPeriod     int64
+	tokensPerSample     int64
+
+	lock            sync.Mutex
+	availableTokens int64
+	lastRefillTime  time.Time
+	nextRefillTime  time.Time
+}
+
+func max(a, b int64) int64 {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+// Sample decides whether to sample a trace. The trace.Sampler type is unfortunately a function and
+// not an interface. This `Sample` function should be provided to the OpenCensus tracing library as
```

Comment:
Just my opinion on API design. Will remove the commentary.

Reply:
As for why "unfortunate," maybe not in this case, but in the more general sense, I find it complicates the ability of test code that needs to be able to control parts of the underlying struct (which I originally thought had been needed for this). Returning the function gives up that possibility since the struct is no longer accessible to anything outside the function including test code.

(Of course, that would only be feasible any way in cases where the code and tests are in the same package which is not the case with this code base.)

Reply:
I think that if it's necessary for tests to modify the struct, then it's more of an indicator that decomposition of your type is needed.